### PR TITLE
refactor: improve type hints

### PR DESCRIPTION
**Pull request by waketzheng**
## Description

Improve type hints
## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added the changelog accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: tortoise/expressions.py [hunk 1]
```diff
@@ -158,7 +147,7 @@ def resolve(self, resolve_context: ResolveContext) -> ResolveResult:
             except KeyError:
                 raise FieldError(
                     f"There is no non-virtual field {self.name} on Model {resolve_context.model.__name__}"
-                )
+                ) from None
```
Comment (henadzit) [inline tortoise/expressions.py:150]: What is it for?

**Review by henadzit — COMMENTED** [review]
A lot of nice improvements! I left a few minor comments.

File: tortoise/__init__.py [hunk 2]
```diff
@@ -488,25 +488,26 @@ async def init(
 
         if config_file:
             config = cls._get_config_from_config_file(config_file)
-
-        if db_url:
+        elif db_url:
             if not modules:
                 raise ConfigurationError('You must specify "db_url" and "modules" together')
             config = generate_config(db_url, modules)
+        else:
```
[the remark this answers is not in the document]
Comment (waketzheng) [reply tortoise/__init__.py:495]: > I think we should just move the `ConfigurationError` here:
> 
> ```python
> elif not config:
>   raise ConfigurationError(
>                 'You should init either from "config", "config_file" or "db_url"'
>             )
> ```
> 
> I think this should fix the typing issue.

In that case, if `int(bool(config) + bool(config_file) + bool(db_url)) > 1` will not raise ConfigurationError

Or do you mean update code like this:
```
-        if int(bool(config) + bool(config_file) + bool(db_url)) != 1:
-            raise ConfigurationError(
-                'You should init either from "config", "config_file" or "db_url"'
-            )

-        if config_file:
+        if config_file and not config and not db_url:
             config = cls._get_config_from_config_file(config_file)
-        elif db_url:
+        elif db_url and not config and not config_file:
             if not modules:
                 raise ConfigurationError('You must specify "db_url" and "modules" together')
             config = generate_config(db_url, modules)
-        else:
-            assert config is not None  # To improve type hints
+        elif not config or int(bool(config) + bool(config_file) + bool(db_url)) > 1:
+            raise ConfigurationError(
+                'You should init either from "config", "config_file" or "db_url"'
+            )
```

File: tortoise/expressions.py (hunk 1 above)
Comment (waketzheng) [reply tortoise/expressions.py:150]: Add `from None` will hide the KeyError message. Now that we should the FieldError, there is no need to show the previous KeyError info.

For Example:
- demo.py
```
class FieldError(Exception):
    pass

try:
    {}["a"]
except KeyError:
    raise FieldError("There is no non-virtual")
```
print:
```
Traceback (most recent call last):
  File "demo.py", line 6, in <module>
    {}["a"]
KeyError: 'a'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "demo.py", line 8, in <module>
    raise FieldError("There is no non-virtual")
__main__.FieldError: There is no non-virtual
```
- demo_with_from_none.py
```py
class FieldError(Exception):
    pass

try:
    {}["a"]
except KeyError:
    raise FieldError("There is no non-virtual") from None
```
print:
```
Traceback (most recent call last):
  File "demo_with_from_none.py", line 8, in <module>
    raise FieldError("There is no non-virtual") from None
__main__.FieldError: There is no non-virtual
```
See more at: https://docs.astral.sh/ruff/rules/raise-without-from-inside-except/

Comment (henadzit) [reply tortoise/expressions.py:150]: Got it, thank you for the explanation!

File: tortoise/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (henadzit) [reply tortoise/__init__.py:495]: I think I misread the conditions of the first check, sorry for the confusion!

**Review by henadzit — APPROVED** [review]

**Pull request merged**